グループ編集機能、メッセージ送信機能の実装

Koh0214 opened the pull request:
# WHAT
### indexビューでdbから取得してきた情報を表示
・rootをmessages controllerからgroups controllerに移した
・上記に伴いビューの場所を適宜変更
・Groups.allの内容をeachを使いサイドバーに表示
・groupsのルートの中にmessagesをネストさせ、idを渡せるようにした

### group_editの実装
・groupsのeditボタンにパスを通した
・editのビューをform_forで書き換え
・collection_check_boxにcheckedのオプションを追加
・updateで保存できるようにrootにupdateを、groups controllerにメソッド追加

### メッセージの送信機能の実装
・messages, indexのビューの下にform_forを設置
・messageに対してcreateのルートを設定
・createメソッドとstrong_parameterの設定(current_user.idを使用)
・caseを使い、bodyに何も入力されていない時に保存できないように設定
・保存されていないときにflashメッセージを表示
・formのレイアウト設定
・メッセージのindexでdbからの情報を表示
・部分テンプレートへの切り分け
・render, collectionを使用した部分テンプレートの表示（https://railsguides.jp/layouts_and_rendering.html）

# WHY

グループの編集機能に必要なため

Commit:
rootをgroups#indexに変更、messagesをネスト

messageをネストし、idを他のビューで使えるようにした。

Commit:
groupsのindexをrootにして、グループの部分のみをサイドバーに繰り返し表示させた

Commit:
groups/indexで使うために@groupsを定義

@groups = Group.allで全部を取ってきている。

Commit:
groups, indexの左側にそれぞれのグループのリンクをeachで表示

Commit:
messages_controllerに@groupsと@groupを設定

Commit:
messages, indexの左側にeachでそれぞれのグループを表示、リンクを繋ぐ

Commit:
form_forへの書き換え（@groupを利用）

Commit:
groups_controllerに対してupdateメソッドを追加、pictweetの記述を参考に追加

Commit:
group内のデータの更新(nameのみ)

Commit:
groups_controllerでuser_idsを保持するように記述、collection_check_boxesのchecked条件で使用

Commit:
user_idsを生のままだとストロングパラメータで弾かれるので、set_params[:user_ids]に変更。

Commit:
updateからredirect_to rootを設定

groups/group.id/messages
にリダイレクトさせたかったが上手くいかなかったのでひとまずrootへ。

Commit:
index画面のCSSのレイアウトを微調整

Commit:
group作成、編集それぞれのCSSを微調整

.chat-group-form__field--leftはもともとwidth 30%だった。

Koh0214 commented:
ルート画面
![2017-01-14 13 58 22](https://cloud.githubusercontent.com/assets/22772196/21952486/b3fcf88c-da61-11e6-8f91-ff81d0c7f28b.png)

ルート画面最下部
![2017-01-14 13 58 31](https://cloud.githubusercontent.com/assets/22772196/21952488/b402cbea-da61-11e6-8c39-c1a2a33f6b70.png)

グループを1つ選択した画面
![2017-01-14 13 58 43](https://cloud.githubusercontent.com/assets/22772196/21952487/b40293be-da61-11e6-84c9-ecaaa71b4d02.png)

新規グループ作成画面
![2017-01-14 13 58 50](https://cloud.githubusercontent.com/assets/22772196/21952490/b40410cc-da61-11e6-8498-b2111daff27b.png)

グループ編集画面
![2017-01-14 13 59 08](https://cloud.githubusercontent.com/assets/22772196/21952489/b403f63c-da61-11e6-9bc4-93da3c604605.png)

Koh0214 commented:
データ編集時のターミナル画面
![2017-01-14 14 02 00](https://cloud.githubusercontent.com/assets/22772196/21952504/2650bb1c-da62-11e6-949b-d05a134fe6d1.png)

db上でのグループ名の変更を確認（group_id: 20)
![2017-01-14 14 02 19](https://cloud.githubusercontent.com/assets/22772196/21952505/26522e52-da62-11e6-8b87-355f675fd3dc.png)

db上でのgroup_userの追加を確認（user_id: 5）
![2017-01-14 14 02 38](https://cloud.githubusercontent.com/assets/22772196/21952506/265569fa-da62-11e6-84b6-e4caae54f04f.png)

Commit:
messagesにcreateのルートを設定

Commit:
インデントの微調整

Commit:
メッセージの部分へform_forでtext_fieldを設置

Commit:
strong_parameterの設定（ムリヤリ）

Koh0214 commented:
![2017-01-14 21 20 24](https://cloud.githubusercontent.com/assets/22772196/21954801/7b89bfce-da9f-11e6-9cf6-c75cdbd3378d.png)

dbを確認(2回saveを押してしまったのでレコードが2つです)
![2017-01-14 21 21 41](https://cloud.githubusercontent.com/assets/22772196/21954802/7b8d8f1e-da9f-11e6-9f3a-40b921fa96d3.png)

Commit:
case文でbodyが無いと保存できないように記述、bodyが無い場合はアラートを出す

Koh0214 commented:
文章の入力がない場合のフラッシュメッセージ
![2017-01-14 22 18 29](https://cloud.githubusercontent.com/assets/22772196/21955120/89cd78b6-daa7-11e6-8352-ebecef77b8d2.png)

Commit:
text_field周りのCSS設定

Koh0214 commented:
フォームのレイアウト修正後
![2017-01-15 12 48 51](https://cloud.githubusercontent.com/assets/22772196/21960114/7b8ae9bc-db21-11e6-8333-43dc96968e2e.png)

madana-i commented on app/assets/stylesheets/modules/_contents.scss:
```diff
@@ -90,15 +92,17 @@
+    font-size: 20px;
+    margin-top: 23px;
     margin-right: 40px;
     padding: 12px 25px 12px 25px;
```
上下　左右　がそれぞれ一致している場合は
padding: 12px 25px;
という風に省略して書けます！

Koh0214 replied:
見落としていました！
他の箇所もあったので、同様に修正しました！

ありがとうございます！

madana-i commented on app/assets/stylesheets/modules/_contents.scss:
```diff
@@ -90,15 +92,17 @@
     color: $darkgrey;
     float: right;
-    margin: 12px 15px 0 0;
+    position: absolute; top: 40px; right: 200px;
```
これは一行にまとめすぎかと思います！
この場合ですと
```css
position: absolute;
top: 40px; right: 200px;
```
という風に書くと見やすいかと思います

Koh0214 replied:
修正しました！
ありがとうございます！

madana-i commented on app/controllers/groups_controller.rb:
```diff
@@ -1,14 +1,26 @@
 class GroupsController < ApplicationController
 
+	def index
```
Rubyのインデントは半角スペース2つですね

Koh0214 replied:
すみません！
修正しました！

sublimeの設定も変更しました！

Koh0214 replied:
修正しました！！

madana-i commented on app/controllers/groups_controller.rb:
```diff
@@ -1,14 +1,26 @@
 class GroupsController < ApplicationController
 
+	def index
+		@groups = Group.all
```
これだと自分に関係のないグループまで取得してしまいませんか？

Koh0214 replied:
groupsのindexビューの中で、全てのグループ名と1つ目のメッセージを表示させているので
.allで取得しています！

Koh0214 replied:
あ、今理解しました！

関係のあるグループのみを取得するように修正します！

madana-i commented on app/controllers/messages_controller.rb:
```diff
@@ -1,4 +1,28 @@
 class MessagesController < ApplicationController
   def index
+  	# binding.pry
```
プルリク出す場合は不要なコメントアウト消しましょう

Koh0214 replied:
申し訳ありません！
修正しました！

madana-i commented on app/controllers/messages_controller.rb:
```diff
@@ -1,4 +1,28 @@
+  end
+
+  def create
+  	case params[:message][:body].empty?
```
true falseで分けるならif文の方が適切ですね

Koh0214 replied:
if 文で修正しました！

caseよりもif文が適切な理由はありますか？
caseは数値などの複数条件の時に使用するのでしょうか？

hand12 replied:
おっしゃるとおりです。case文が、複数の条件に対しての条件分岐に対して、if文はtrueとfalseの2択の条件分岐に使用します。
今回は保存できた時、できなかった時の2択なのでifを使いましょう！

madana-i commented on app/controllers/messages_controller.rb:
```diff
@@ -1,4 +1,28 @@
+  def create
+  	case params[:message][:body].empty?
+  	when true
+  		# window.alert("何か書かないと送信できません")
```
コメントアウト消しましょう

Koh0214 replied:
修正しました！

madana-i commented on app/controllers/messages_controller.rb:
```diff
@@ -1,4 +1,28 @@
+  		# flash.now[:notice] = '何か書かないと送信できません'
+  	when false
+	  	message = Message.create(set_params)
+	  	redirect_to request.original_url
```
prefixで指定するようにしましょう

Koh0214 replied:
修正しました！

madana-i commented on app/controllers/messages_controller.rb:
```diff
@@ -1,4 +1,28 @@
+
+  private
+  def set_params
+  	params[:user_id] = current_user.id
```
user_idはアソシエーション使用して入れるようにしましょう

madana-i commented on config/routes.rb:
```diff
@@ -1,6 +1,7 @@
+  root 'groups#index'
+  resources :groups, only: [:new, :create, :edit, :update] do
+	  resources :messages, only: [:index, :create]
+	end
```
インデントずれてますね

Koh0214 replied:
タブとスペースが原因だったようです！

修正しました！

Commit:
インデントの修正(tab, space)

Commit:
インデントの修正

Commit:
インデントをtab => spaceに修正

Commit:
paddingの指定の仕方の修正

Commit:
いらないコメントアウトの削除

Commit:
条件分岐をif文で書き換え、redirectをprefixを使い書き換え

Commit:
アソシエーションを使ってリファクタリング

@groups = current_user.groups
で、current_userの持つgroupsを取ってくる

message = current_user.messages.create(set_params)

でcurrent_userのidを入れたmessageをcreateする。

set_paramsの中身は
[:message[:body]]と、[group_id: params[:group_id]]の2つ。
これにcurrent_userから来たidが加わる。

hand12 commented on app/assets/stylesheets/modules/_contents.scss:
```diff
@@ -90,15 +92,17 @@
+    font-size: 20px;
+    margin-top: 23px;
     margin-right: 40px;
     padding: 12px 25px 12px 25px;
```
ここpushしわすれていませんか？

Koh0214 replied:
pushしました！

hand12 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,32 +1,32 @@
-		@groups = Group.all
-	end
+  def index
+    @groups = Group.all
```
ここの変更がまだできていないので修正しましょう！

Koh0214 replied:
修正しました！

hand12 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,32 +1,32 @@
-		redirect_to :root
-	end
+  def create
+    Group.create(set_params)
```
createを使う際は条件分岐もしましょう！

hand12 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,32 +1,32 @@
-	end
+  def update
+    group = Group.find(params[:id])
+    group.update(set_params)
```
こちらもですね！

hand12 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,28 +1,22 @@
-	  	redirect_to request.original_url
-	  end
+    if params[:message][:body].present?
+      message = current_user.messages.create(set_params)
```
Railsの条件分岐をnewとsaveを使って考えてみてください！！

hand12 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,28 +1,22 @@
+      message = current_user.messages.create(set_params)
+      redirect_to group_messages_path
+    else
+      redirect_to group_messages_path, notice: '何か書かないと送信できません'
```
redirect_toとrenderの違いを調べてどちらがいいかを考えてみてください！

Koh0214 replied:
renderを使う。
・インスタンス変数の保持
・速度
の2つがメリット

renderはコントローラーを介さずにビューを呼ぶので、
一旦生成したインスタンス変数が更新されない。

ex)
グループ名とユーザー名を入れる時、
どちらもないとダメと設定して、
ユーザー名がなくて弾かれた時、
renderを使って入ればグループ名は保持されたままになる。


redirect_toはインスタンス変数が新しく生成される。

hand12 reviewed (commented):
コメントしました！

Commit:
groups_controllerの@groupsを修正

Commit:
messageを保存できるかで条件分岐、notice => alertに修正

Commit:
サインインしているユーザーのグループのみ表示、セーブできるか分岐、new&create, edit&updateでまとめた

Commit:
group_updateの条件分岐

hand12 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,27 +1,38 @@
 
   def index
-    @groups = Group.all
+    if user_signed_in?
```
ここはbefore_actionで書いたほうがよさそうです！
今はindexとcreateくらいしかアクションはないのですが、どちらもログインしていないと動いて欲しくないアクションかつ、これからもアクションは増える可能性もあるので、before_actionで定義してあげましょう！

Koh0214 replied:
before_action定義しました！

便利です！

hand12 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,27 +1,38 @@
+  def create 
+    group = Group.new(set_params)
+    if group.save
+      @user_ids = set_params[:user_ids]
```
この記述入らないですね！

hand12 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,27 +1,38 @@
-    @user_ids = set_params[:user_ids]
-    redirect_to :root
+  def edit
+    @group = Group.find(params[:id])
```
updateでも同じ記述がありますので、こちらもメソッドにまとめてみましょう！

Koh0214 replied:
メソッドにまとめました！！

Commit:
必要のない変数、オプションの削除

hand12 commented on app/controllers/messages_controller.rb:
```diff
@@ -7,10 +7,14 @@ def index
 
   def create
     if params[:message][:body].present?
```
ここの記述は、params[:message]の中にbodyというキーがあるかどうかをチェックしていると思うのですが、モデルbのvalidationの中にすでにその記述は書いてあるはずなので、ここで条件分岐する必要はないですね！

Koh0214 replied:
削除しました！

hand12 reviewed (commented):
コメントしました！

Commit:
ログインしていない場合のbefore_actionを定義

Commit:
before_actionにflashメッセージを設定

Commit:
bodyが存在するかどうかのvalidationを削除

Commit:
DRYのために記述をメソッドにまとめた。アラート文の微調整

hand12 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,20 +1,46 @@
+  end
 
+  def edit
+    @group = find_group_id
```
Railsではこのような、findメソッドを使って送られてきたparamsのレコードを取得する際は、set_groupというようなメソッドを作成して、そのメソッドをbefore_actionで呼ぶことが一般的です。
今回は@が付いている時と付いていない時で分けた方のかもしれないですが、それよりもRailsの定石に則って、before_actionで定義したほうがよさそうです。

Koh0214 replied:
set_groupで定義しました！！

hand12 commented on app/views/groups/edit.html.haml:
```diff
@@ -1,25 +1,24 @@
-    %input{ name: "utf8", type: "hidden", value: "✓" }/
-    %input{ name: "_method", type: "hidden", value: "patch" }/
-    %input{ name: "authenticity_token", type: "hidden", value: "U5ABp0/UobRQTSwsR0V40bX1w3t0rZ0CTqrVyLecD+L6sdhBRO/gVcKy7w/Rgxnc7Hf0Ts39996dal3to5fTag==" }/
+  = form_for @group, html: { id: 'edit_chat_group_22', class: 'edit_chat_group' } do |f|
```
このidやclass、もし使っていないのであれば削除してしまいましょう！

Koh0214 replied:
削除しました！

hand12 commented on app/views/groups/index.html.haml:
```diff
@@ -0,0 +1,15 @@
+          %i.fa.fa-cog{ aria: { hidden: true } }
+      %h5.user_name kohei
+    .left__content
+      - @groups.each do |group| 
```
eachは、部分テンプレートに直すようにしましょう！

Koh0214 replied:
部分テンプレートの方へ移しました！

Koh0214 replied:
複数箇所から部分テンプレートを使うと、便利さがすごくわかります！！

hand12 commented on app/views/messages/index.html.haml:
```diff
@@ -8,21 +8,15 @@
-      %p.group_name_small koheiとゆかいな仲間たち
-      %p.group_message やっほおおおおお
-
+      - @groups.each do |group| 
```
こちらもですね！

Koh0214 replied:
部分テンプレートへ移しました！

hand12 commented on app/views/messages/index.html.haml:
```diff
@@ -85,10 +79,9 @@
-        .photo_icon
-          %i.fa.fa-picture-o{ aria: { hidden: true } }
-        %p.box_message type your message, yeah
+      = form_for @message, :url => {:action => :create} do |f|
```
このurlなども入らなければ消しましょう！

Koh0214 replied:
ネストしたルートに飛ばすためこの記述を入れていました。

form_forに対応した記述に変更しました！

hand12 reviewed (commented):
コメントしました！

Commit:
要らないid, classの削除

Commit:
form＿forをネストしたルートの記述方法に変更

Commit:
仮置きのhtmlから、データベースの情報を参照するように修正

Commit:
部分テンプレートの作成と適用

Commit:
set_groupの作成

基本的に、@groupを毎アクションセットする。

indexの時はどのグループというのがなく、group_idがなくなるのでエラーが出る。
なのでexcept: : index

Commit:
インデントの微調整

Commit:
部分テンプレートをsharedディレクトリに移して、rootからも使用

Commit:
部分テンプレートの仕組み理解のために変数名を変更

Commit:
chatエリアの背景CSS微調整

Koh0214 commented:
### 各種情報はdbから取得したもの。
#### 左カラム
現在のユーザー名、各グループ名、グループ名下の最新メッセージ

#### 右カラム
グループ名、メンバー一覧、投稿者名、投稿時間、投稿内容

![2017-01-16 22 15 32](https://cloud.githubusercontent.com/assets/22772196/21984455/5b3f8ae6-dc39-11e6-9488-88167e898062.png)

koheiiwamura commented on app/assets/stylesheets/modules/_contents.scss:
```diff
@@ -14,11 +14,13 @@
   .group_name_small{
+    color: white;
     font-size: 15px;
     padding: 20px 0 5px 0;
```
短くできますね！

Koh0214 replied:
CSS全体をリファクタリングしました！
ありがとうございます！

koheiiwamura commented on app/assets/stylesheets/modules/_contents.scss:
```diff
@@ -90,15 +92,18 @@
     float: right;
-    margin: 12px 15px 0 0;
+    position: absolute;
+    top: 40px; right: 200px;
```
rightのところは改行しましょう

Koh0214 replied:
改行しました！

koheiiwamura commented on app/controllers/groups_controller.rb:
```diff
@@ -1,20 +1,49 @@
+
+  def find_group_id
+    Group.find(params[:id])
+  end
```
このメソッドはどこでつかってますか？

Koh0214 replied:
見落としていました！削除します！

koheiiwamura commented on app/views/shared/_message.html.haml:
```diff
@@ -0,0 +1,6 @@
+%p.timestamp
+  = message.created_at
+%p.post
+  = message.body
```
最後は改行しましょう。

Koh0214 replied:
すみません！改行しました！

koheiiwamura commented on app/views/shared/_group.html.haml:
```diff
@@ -0,0 +1,5 @@
+  %p.group_name_small 
+    = group.name
+  %p.group_message
+    = group.messages.last[:body]
```
最後は改行で。

Koh0214 replied:
改行しました！

Commit:
使っていないメソッドの削除

Commit:
改行の追加

Commit:
CSSの微調整

Commit:
CSSのリファクタリング

koheiiwamura commented on app/views/messages/index.html.haml:
```diff
@@ -6,89 +6,25 @@
-      %p.group_message やっほおおおおお
-
+      - @groups.each do |partial_group| 
+        = render partial: "shared/group", locals: { group: partial_group }
```
これeachで部分テンプレートを繰り返すのではなくて、collectionなどをつかって簡潔に書いてみましょう。

Koh0214 replied:
render, collectionを使い書き換えました！

koheiiwamura commented on app/views/messages/index.html.haml:
```diff
@@ -6,89 +6,25 @@
-      %p.timestamp 1993/02/14 07:32:14
-      %p.post にゃおん。
+      - @group.messages.each do |partial_message|
+        = render partial: "shared/message", locals: { message: partial_message }
```
ここもを使わずに簡潔にかけます。

Koh0214 replied:
render, collectionを使い書き換えました！
ありがとうございます！

Commit:
renderを使い、部分テンプレートを書き換え
